tests/scripts/audit-validity-dates.py

```diff
@@ -0,0 +1,411 @@
+#!/usr/bin/env python3
+#
+# copyright the mbed tls contributors
+# spdx-license-identifier: apache-2.0
+#
+# licensed under the apache license, version 2.0 (the "license"); you may
+# not use this file except in compliance with the license.
+# you may obtain a copy of the license at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Audit validity date of X509 crt/crl/csr
+
+This script is used to audit the validity date of crt/crl/csr used for testing.
+The files are in tests/data_files/ while some data are in test suites data in
+tests/suites/*.data files.
+"""
+
+import os
+import sys
+import re
+import typing
+import types
+import argparse
+import datetime
+import glob
+from enum import Enum
+
+from cryptography import x509
```

Comment:
Which version(s) of Python and the cryptography package have you tested with? I get errors with all versions of Ubuntu using the system-provided `python3-cryptography`.

* Ubuntu 16.04 (Python 3.5.1, cryptography 1.2.3):
    ```
  File "./tests/scripts/audit-validity-dates.py", line 55
        self.not_valid_after: datetime.datetime
                            ^
    SyntaxError: invalid syntax
    ```
* Ubuntu 18.04 (Python 3.6.7, cryptography 2.1.4):
    ```
    Traceback (most recent call last):
      File "./tests/scripts/audit-validity-dates.py", line 411, in <module>
        main()
      File "./tests/scripts/audit-validity-dates.py", line 401, in main
        td_auditor.walk_all(data_files)
      File "./tests/scripts/audit-validity-dates.py", line 227, in walk_all
        data_list = self.parse_file(filename)
      File "./tests/scripts/audit-validity-dates.py", line 200, in parse_file
        result = self.parse_bytes(data)
      File "./tests/scripts/audit-validity-dates.py", line 210, in parse_bytes
        result = self.parser[data_type](data)
      File "./tests/scripts/audit-validity-dates.py", line 106, in wrapper
        result = der_loader(data)
    TypeError: load_der_x509_certificate() missing 1 required positional argument: 'backend'
    ```
* Ubuntu 20.04 (Python 3.8.2, cryptography 2.8): same as 18.04
* Ubuntu 22.04 (Python 3.10.6, cryptography 3.4.8):
    ```
    Traceback (most recent call last):
      File "/home/gilpes01/z/dev/mbedtls/reviewing/7399-issue%7014%certificate-audit-script/./tests/scripts/audit-validity-dates.py", line 411, in <module>
        main()
      File "/home/gilpes01/z/dev/mbedtls/reviewing/7399-issue%7014%certificate-audit-script/./tests/scripts/audit-validity-dates.py", line 402, in main
        sd_auditor.walk_all(suite_data_files)
      File "/home/gilpes01/z/dev/mbedtls/reviewing/7399-issue%7014%certificate-audit-script/./tests/scripts/audit-validity-dates.py", line 227, in walk_all
        data_list = self.parse_file(filename)
      File "/home/gilpes01/z/dev/mbedtls/reviewing/7399-issue%7014%certificate-audit-script/./tests/scripts/audit-validity-dates.py", line 355, in parse_file
        audit_data = self.parse_bytes(bytes.fromhex(match.group('data')))
      File "/home/gilpes01/z/dev/mbedtls/reviewing/7399-issue%7014%certificate-audit-script/./tests/scripts/audit-validity-dates.py", line 216, in parse_bytes
        audit_data.fill_validity_duration(result)
      File "/home/gilpes01/z/dev/mbedtls/reviewing/7399-issue%7014%certificate-audit-script/./tests/scripts/audit-validity-dates.py", line 68, in fill_validity_duration
        self.not_valid_after = x509_obj.next_update
      File "/usr/lib/python3/dist-packages/cryptography/hazmat/backends/openssl/x509.py", line 289, in next_update
        self._backend.openssl_assert(nu != self._backend._ffi.NULL)
      File "/usr/lib/python3/dist-packages/cryptography/hazmat/backends/openssl/backend.py", line 242, in openssl_assert
        return binding._openssl_assert(self._lib, ok, errors=errors)
      File "/usr/lib/python3/dist-packages/cryptography/hazmat/bindings/openssl/binding.py", line 77, in _openssl_assert
        raise InternalError(
    cryptography.exceptions.InternalError: Unknown OpenSSL error. This error is commonly encountered when another library is not cleaning up the OpenSSL error stack. If you are using cryptography with another library that uses OpenSSL try disabling it before reporting a bug. Otherwise please file an issue at https://github.com/pyca/cryptography/issues with information on how to reproduce this. ([])
    ```

Under 22.04, the script works with some files. I guess the internal error comes from malformed files but I haven't investigated.

Due to the state of our CI, we currently need the script to be syntactically valid and pass pylint and mypy with Python 3.5. It's ok if it requires a more recent version to work usefully, although it would be nice to support Ubuntu 20.4 since many people on the team haven't upgraded yet.

Reply:
I'm using Python 3.8.0 and cryptography 40.0.1. I think updating cryptography could make the script useful.

Reply:
> Under 22.04, the script works with some files. I guess the internal error comes from malformed files but I haven't investigated.

Yes, it seems that `cryptography 3.4.8` raise `InternalError` when there are malformed data, which is different from `cryptography 3.4.8`.

> although it would be nice to support Ubuntu 20.4 since many people on the team haven't upgraded yet.

The earliest version of cryptography that is compatible with this script should be `35.0.0`, and it should available for `Python 3.8.2` on Ubuntu 20.04.